### DISCUSSION TITLE
Optional param casted automatically on routerLink even if we don't want optional param (when param is already set in url)

### DISCUSSION THREAD

**Discussion by JulienMalcouronne** [vuejs/router#2247] [Help and Questions] [answered]
Hello I'm not sure this is a bug or perhaps there are wrong stuff in our codebase. 

Once we set a param to be optional in the routes definition, the ```RouterLink :to={...route.name}``` without parameters in the code then the param is automatically set if we currently are on this route. 

Is there a way to prevent that to happening?

**Example:** 
for instance if I'm on the following route ```{name: dashboard, params: {id: 1}}```

I have the following links -> 

``` {name: dashboard}``` -> the first ```<a href="dashboard/1">``` where it should be only /dashboard
``` {name: dashboard, params: {id: 2}}```

**We are using:** 
 "vue-router": "^4.3.2",
 "vue": "3.4.27",

**Comment by skirtles-code** [comment]
I attempted to reproduce the problem you're seeing based on the examples you gave, but it seemed to be working fine for me:

- [Playground](<https://play.vuejs.org/#eNq1U8lu2zAQ/RVCLSAHsMVmQQ6CkqZNgy7oEqRFL1UPjEXbTCSS4OK4MPTvHZEUZceJgR56MExxZt68efO4ThrCeHankzxhjRTKoDWaKkoMfSMlatFMiQalS0vTkocEJayhKkQy7D+HcFfXx4iUma8teUQdwe+g5AhlVtORL/ffjbDcjNIXUJYeJOMklAO3wtBG1lB/3iUW0v3BQRsl+Pz80ipFeaCGJDGLvMAhhtZr9NIFspmt62sIorZ1MNjjFLbu8WoWTnC+ccw+M36PciPOymSNOGlojtKK6MWtIKpKUVsm5+8FMgLheFvgobQHxhH5n3uMYSBFGp3DaliVo8OO/m5XfPif+x490/doX98CB3GLzmh91Of/ZPQB4ZAWwgWOmwYDeHM8ac8vtBHqzwemDfyNw6XH3bTtJNozGvRdz9z17626dTuYVnAdbKXRGfrVcY1SLURDnUpgN5RiOE4FtODgxPxRl3a8Vbktsi8fBM1Z9XovWMl/PyKngNymBKN112/h1cmfkmx04Cj50UrewgssOV05hSo6I7buoWEPO+LsPsmKLcN2o2viu/M38dX51O1N+91MGiJh24IDvpugDAFdJgDo8ctk2Gt3XSYLY6TOMbZc3s8z0A0PGRcn2XH2Cnpqs3GbUd1MbpV40M5eZeLE6LAvIAlXdGmEqPWESPZci53Ei9PsNDsE499iQMeMV3TlsDtoGL2FMY2Gpc3Y/NGQ3apZTdU3aRgsdWtYUtfi4ZO7M8rSSHS6oNP7J+7v9MpTvlYUplvSjeEMUXNqfPjq+1e6gnMMNqKyNWTvCd5QLWrbcfRpby2vgPZGnmP70e2M8fkPfbUylOt+qI6oU8Plu0Ve7hl9oHucnUQV27/hwj7r>)

I didn't really understand what you meant about `RouterLink :to={...route.name}`. Perhaps you wanted something like this?

```html
<RouterLink :to="{ ...$route, params: {} }">
```

**Reply by JulienMalcouronne** [reply]
Hi Skirtles, 

Thanks a lot for giving it a shot. 

Actually I succeed to reproduce the issue. 

I have updated your playground with more nested routes. Indeed it is working as expected when there are only one optional param in params. 

If you try to reach different routes in different order then the params are not as expected.

To display it a more obvious way, I have destructured the href. 

[Playground](https://play.vuejs.org/#eNrlVm1v2zYQ/iuENkAOYItL2vSDoGbZumLrsJciG/Zl6gfGom02FCmQlOPB0H/f8UWUFNtZBxRokQYISN0d7+55eEffPqkJE9l7neQJqxupDNqjpaLE0O+aBnVopWSN0m1L01IEAyVbQ1XQZNh/Dmp7rteRpsn82VJErzP4PysFQlmr6cwf99+1bIWZpV/BsfQsmSfhOORWGFo3HM5fWcOicQtstFFSrK9etUpREVJDDTGbvMBBh/Z79LVTZKuW87egRF2HnB/sHRUtDw6Rk4OEsygpblyKvzBxh7YLzaV5WSb7jaKrORJky9aQVlcmKDfSKkBW0xylFdGbW0lUlSLQXv0okZEIR6mlxPuPMND1krPlHTjp3cK5G0o44ja2jZijgqDc7sDKLmAB+OwOQBWYXEXcMX085N/ThiO8jwp0DtQrUusciohVOTqHnI5Ax+dfAviLE+AvPgPwHw2+VGsimCaGSXHAQAoSMHgD28tTZOCxh8svhpqL/8kMnlDzRBiC51/pR6hJ4eW03AzA/4sl5zE84U+HoaP1cYh7VB9PvTJC03x4YVz2FI2b6BOyVOAwchR2/ur1/sRfjN4j7LU4qAscByCYi/zMdHRq+5XWUv3zE9MGlnkQer/jaW4Rp7Y4t/3Qc+fi9xPcRDrMclLoMG1p9BL9bXONl7eRNXV3B1MYSjFslxJCCBjQ8gdRuvnk5PjH1FPSOxkuNmfVt1F9ynHUbxivYDLMfYr279T7PBRGH3NSQLmrvCHyh0Q/kcEki77Sp497n0LQTpUnw07Nxl+d/3jnN7CA7bsHF6ngIsflMttb242vpPxYec3OXEhfBnlYS9HBMF8KunNVVdEVaXkfAmr3oKAOp/uKbUNHxP6PI7yX2PZ37eFNp93h63lRkwY6RArw75CUQaHLBBx6/2Uy9IIVQ3Mb0+gc41Y0d+sMqMaDxfXz7Fn2DcTUZiTNqK4Xt0rew11BwDIJ91Am12CEK7o1UnK9IA07FeLA8PpF9iI7h+fiFoN3zERFd863dQ3QO4BpNFzeiq0fgLTVwThVvze2bqdgCefy/mcnM6qlMdHlhi7vjsjf651P+a2igG4LD2PUGaLW1Hj16z9+ozvYR2Utq5aD9SPKG6olb22O3uz7VlSQ9sjOZfvG3RkT6z/1652hQvegbKKODWfvLvLVI9CHdJ9lzyOL3b/DndrT) 

Could you please let me know if you think there is an issue on Vue-Router ?

**Reply by posva** [reply]
Your routes have an extra leading slash. See the routes in [Playground](https://play.vuejs.org/#eNrlVm1v2zYQ/iuENkAOYItL2vSDoGbZumLrsJciG/Zl6gfGom02FCmQlOPB0H/f8UWUFNtZBxRokQYISN0d7+55eEffPqkJE9l7neQJqxupDNqjpaLE0O+aBnVopWSN0m1L01IEAyVbQ1XQZNh/Dmp7rteRpsn82VJErzP4PysFQlmr6cwf99+1bIWZpV/BsfQsmSfhOORWGFo3HM5fWcOicQtstFFSrK9etUpREVJDDTGbvMBBh/Z79LVTZKuW87egRF2HnB/sHRUtDw6Rk4OEsygpblyKvzBxh7YLzaV5WSb7jaKrORJky9aQVlcmKDfSKkBW0xylFdGbW0lUlSLQXv0okZEIR6mlxPuPMND1krPlHTjp3cK5G0o44ja2jZijgqDc7sDKLmAB+OwOQBWYXEXcMX085N/ThiO8jwp0DtQrUusciohVOTqHnI5Ax+dfAviLE+AvPgPwHw2+VGsimCaGSXHAQAoSMHgD28tTZOCxh8svhpqL/8kMnlDzRBiC51/pR6hJ4eW03AzA/4sl5zE84U+HoaP1cYh7VB9PvTJC03x4YVz2FI2b6BOyVOAwchR2/ur1/sRfjN4j7LU4qAscByCYi/zMdHRq+5XWUv3zE9MGlnkQer/jaW4Rp7Y4t/3Qc+fi9xPcRDrMclLoMG1p9BL9bXONl7eRNXV3B1MYSjFslxJCCBjQ8gdRuvnk5PjH1FPSOxkuNmfVt1F9ynHUbxivYDLMfYr279T7PBRGiDmpn9wV3hD4Q4KfSGCSRF/o07c9ZBCUU93JqFOz8VfnP975DSxg++7BNSq4xnGxzPbWduPrKD9WXLMzF9IXQR7WUnQwypeC7lxNVXRFWt6HgMo9KKfD2b5i29APsfvjAO8ltvldc3jTaW/4al7UpIH+kAL8OyRlUOgyAYfef5kMnWDF0NrGNDrHuBXN3ToDqvFgcf08e5Z9AzG1GUkzquvFrZL3cFcQsEzCPZTJNRjhim6NlFwvSMNOhTgwvH6RvcjO4bG4xeAdM1HRnfNtXQP0DmAaDZe3YusHIG11ME7V740t2ylYwrm8/9nJjGppTHS5ocu7I/L3eudTfqsooNvCsxh1hqg1NV79+o/f6A72UVnLquVg/YjyhmrJW5ujN/u+FRWkPbJz2b5xd8bE+k/9emeo0D0om6hjw9m7i3z1CPQh3WfZ88hi9y8Y4Np1)

**Reply by JulienMalcouronne** [reply]
Indeed you're right sorry for that. Then I don't know why we have such behavior for instance if you take a look at the following screenshots: 

![Screenshot 2024-05-27 at 09 32 34](https://github.com/vuejs/router/assets/87890111/53b20969-57f9-4fb5-a627-f165c73ca827)
![Screenshot 2024-05-27 at 09 34 31](https://github.com/vuejs/router/assets/87890111/661178b1-04e9-4fd2-b69d-d77ae682918d)

This is the optional param setup:

![Screenshot 2024-05-27 at 09 37 13](https://github.com/vuejs/router/assets/87890111/e082b652-1107-4307-b77f-25fcae5d7447)


I've investigate the source code of Vue-router, was planning to add some debugger / console.log to debug and understand locally, if there is something wrong it must be in router.resolve(), noticed it was used in RouterLink setup function.

**Reply by posva** [reply]
I think you want to add what skirtles said: `params: {}`.

FYI: I can [offer help over a call](https://cal.com/posva/consultancy)

**Reply by JulienMalcouronne** [reply]
> I think you want to add what skirtles said: `params: {}`.
> 
> FYI: I can [offer help over a call](https://cal.com/posva/consultancy)

Thanks a lot for your interest this might work in case I've got only one optional param, but what I did show you was the simplest example we've got. 

I can't pass ```params: {}``` because in my scenario I need the orgId param. I want to provide something like this: 

![Screenshot 2024-05-27 at 11 00 06](https://github.com/vuejs/router/assets/87890111/928949af-7905-44fb-8f81-c372e390ec36)

![Screenshot 2024-05-27 at 11 00 41](https://github.com/vuejs/router/assets/87890111/441c21ff-fe30-4b0b-a6b9-ef9450653e2d)


Well in my case I will simply change the first url to be the parent route and manage the redirection in beforeRouteUpdate but I really think there is something wrong in the router.resolve function related to nested routes optional params.

**Reply by JulienMalcouronne** [reply]
Also what you did suggest doesn't seem to work :( 

I wish it would
![Screenshot 2024-05-27 at 11 08 12](https://github.com/vuejs/router/assets/87890111/bf219b3f-d53e-42d2-b648-187c939edba1)

**Reply by skirtles-code** [reply, marked as answer]
I can't really tell what I'm supposed to be looking at in those screenshots.

Perhaps this is what you're looking for?

https://router.vuejs.org/guide/essentials/navigation.html#Navigate-to-a-different-location

> For optional params, you can provide an empty string (`""`) or `null` as the value to remove it.

**Reply by JulienMalcouronne** [reply]
Thanks a lot for your time found an acceptable workaround. 

I don't have enough time for now on, but later next month I'll try to dig deeper in Vue-router source code to either open an issue or contribute to the project with a pull request to fix that error. 

I still think there is an inheritance problem with optional params when we use the "name" props on RouterLink, that's a bit sad to do programmatic push while we could simply use a link.  

About the screenshots, we can just see on the bottom left that the link's href is wrong (when we hover the link) despite the fact that the code looks okay (even your solution passing params: {}, or params: {orgId: null} wasn't successful).

But in all case I would like to thank you again for your time & investigation.
